File: hashing/maximum_sum.java
public int getDigitSum(int num)
{
    int digitSum = 0;
    while (num > 0)
    {
        digitSum += num % 10;
        num /= 10;
    }

    return digitSum;
}

public int maximumSum1(int[] nums)
{
    Map<Integer, List<Integer>> dic = new HashMap<>();
    for (int num : nums)
    {
        int digitSum = getDigitSum(num);
        if (!dic.containsKey(digitSum))
        {
            dic.put(digitSum, new ArrayList<Integer>());
        }
        dic.get(digitSum).add(num);
    }

    int ans = -1;
    for (int key : dic.keySet())
    {
        List<Integer> curr = dic.get(key);
        if (curr.size() > 1)
        {
            Collections.sort(curr, Collections.reverseOrder());
            ans = Math.max(ans, curr.get(0) + curr.get(1));
        }
    }

    return ans;
}

public int maximumSum2(int[] nums)
{
    Map<Integer, Integer> dic = new HashMap<>();

    int ans = -1;

    for (int num : nums)
    {
        int digitSum = getDigitSum(num);
        if (dic.containsKey(digitSum))
        {
            ans = Math.max(ans, num + dic.get(digitSum));
        }
        dic.put(digitSum, Math.max(dic.getOrDefault(digitSum, 0), num));
    }

    return ans;
}

File: hashing/maximum_sum.py
from collections import defaultdict

def get_digit_sum(num: int):
    digit_sum = 0
    while num:
        digit_sum += num % 10
        num //= 10
    
    return digit_sum

def maximumSum1(nums: List[int]) -> int:
    dic = defaultdict(list)
    for num in nums:
        digit_sum = get_digit_sum(num)
        dic[digit_sum].append(num)
    
    ans = -1
    for key in dic:
        curr = dic[key]
        if len(curr) > 1:
            curr.sort(reverse=True)
            ans = max(ans, curr[0] + curr[1])
    
    return ans

def maximumSum2(nums: List[int]) -> int:
    dic = defaultdict(int)
    ans = -1

    for num in nums:
        digit_sum = get_digit_sum(num)
        if digit_sum in dic:
            ans = max(ans, num + dic[digit_sum])
        
        dic[digit_sum] = max(dic[digit_sum], num)
    
    return ans
